src/main/java/rx/observables/AsyncOnSubscribe.java

```diff
@@ -0,0 +1,510 @@
+                onUnsubscribe.call();
+            }};
+        return new AsyncOnSubscribeImpl<Void, T>(nextFunc, wrappedOnUnsubscribe);
+    }
+
+    /**
+     * An implementation of AsyncOnSubscribe that delegates
+     * {@link AsyncOnSubscribe#next(Object, long, Observer)},
+     * {@link AsyncOnSubscribe#generateState()}, and {@link AsyncOnSubscribe#onUnsubscribe(Object)}
+     * to provided functions/closures.
+     *
+     * @param <S>
+     *            the type of the user-defined state
+     * @param <T>
+     *            the type of compatible Subscribers
+     */
+    private static final class AsyncOnSubscribeImpl<S, T> extends AsyncOnSubscribe<S, T> {
+        private final Func0<? extends S> generator;
+        private final Func3<? super S, Long, ? super Observer<Observable<? extends T>>, ? extends S> next;
+        private final Action1<? super S> onUnsubscribe;
+
+        private AsyncOnSubscribeImpl(Func0<? extends S> generator, Func3<? super S, Long, ? super Observer<Observable<? extends T>>, ? extends S> next, Action1<? super S> onUnsubscribe) {
+            this.generator = generator;
+            this.next = next;
+            this.onUnsubscribe = onUnsubscribe;
+        }
+
+        public AsyncOnSubscribeImpl(Func0<? extends S> generator, Func3<? super S, Long, ? super Observer<Observable<? extends T>>, ? extends S> next) {
+            this(generator, next, null);
+        }
+
+        public AsyncOnSubscribeImpl(Func3<S, Long, Observer<Observable<? extends T>>, S> next, Action1<? super S> onUnsubscribe) {
+            this(null, next, onUnsubscribe);
+        }
+
+        public AsyncOnSubscribeImpl(Func3<S, Long, Observer<Observable<? extends T>>, S> nextFunc) {
+            this(null, nextFunc, null);
+        }
+
+        @Override
+        protected S generateState() {
+            return generator == null ? null : generator.call();
+        }
+
+        @Override
+        protected S next(S state, long requested, Observer<Observable<? extends T>> observer) {
+            return next.call(state, requested, observer);
+        }
+
+        @Override
+        protected void onUnsubscribe(S state) {
+            if (onUnsubscribe != null)
+                onUnsubscribe.call(state);
+        }
+    }
+
+    @Override
+    public final void call(Subscriber<? super T> actualSubscriber) {
+        S state = generateState();
+        UnicastSubject<Observable<T>> subject = UnicastSubject.<Observable<T>> create();
+        AsyncOuterSubscriber<S, T> outerSubscriberProducer = new AsyncOuterSubscriber<S, T>(this, state, subject);
+        actualSubscriber.add(outerSubscriberProducer);
+        Observable.concat(subject).unsafeSubscribe(Subscribers.wrap(actualSubscriber));
+        actualSubscriber.setProducer(outerSubscriberProducer);
+    }
+
+    private static class AsyncOuterSubscriber<S, T> extends ConcurrentLinkedQueue<Long>implements Producer, Subscription, Observer<Observable<? extends T>> {
```

Comment:
`ConcurrentLinkedQueue` seems to be expensive for the usage pattern; you could get away with `SpscLinkedQueue` or `SpscLinkedArrayQueue` here (no need for `MpscLinkedQueue` because the offer side is inside a synchronized block).